lib/CrossTU/CrossTranslationUnit.cpp

```diff
@@ -338,33 +432,63 @@ llvm::Expected<ASTUnit *> CrossTranslationUnitContext::loadExternalAST(
       ++NumNotInOtherTU;
       return llvm::make_error<IndexError>(index_error_code::missing_definition);
     }
-    StringRef ASTFileName = It->second;
-    auto ASTCacheEntry = FileASTUnitMap.find(ASTFileName);
+    StringRef ASTSource = It->second;
+    auto ASTCacheEntry = FileASTUnitMap.find(ASTSource);
     if (ASTCacheEntry == FileASTUnitMap.end()) {
-      IntrusiveRefCntPtr<DiagnosticOptions> DiagOpts = new DiagnosticOptions();
-      TextDiagnosticPrinter *DiagClient =
-          new TextDiagnosticPrinter(llvm::errs(), &*DiagOpts);
-      IntrusiveRefCntPtr<DiagnosticIDs> DiagID(new DiagnosticIDs());
-      IntrusiveRefCntPtr<DiagnosticsEngine> Diags(
-          new DiagnosticsEngine(DiagID, &*DiagOpts, DiagClient));
-
-      std::unique_ptr<ASTUnit> LoadedUnit(ASTUnit::LoadFromASTFile(
-          ASTFileName, CI.getPCHContainerOperations()->getRawReader(),
-          ASTUnit::LoadEverything, Diags, CI.getFileSystemOpts()));
-      Unit = LoadedUnit.get();
-      FileASTUnitMap[ASTFileName] = std::move(LoadedUnit);
+      if (!OnDemandParsingDatabase) {
+        // If no \p OnDemandParsingDatabase is given, try to load from AST dump
+        // file, as on-demand parsing is disabled.
+        IntrusiveRefCntPtr<DiagnosticOptions> DiagOpts =
+            new DiagnosticOptions();
+        TextDiagnosticPrinter *DiagClient =
+            new TextDiagnosticPrinter(llvm::errs(), &*DiagOpts);
+        IntrusiveRefCntPtr<DiagnosticIDs> DiagID(new DiagnosticIDs());
+        IntrusiveRefCntPtr<DiagnosticsEngine> Diags(
+            new DiagnosticsEngine(DiagID, &*DiagOpts, DiagClient));
+
+        std::unique_ptr<ASTUnit> LoadedUnit(ASTUnit::LoadFromASTFile(
+            ASTSource, CI.getPCHContainerOperations()->getRawReader(),
+            ASTUnit::LoadEverything, Diags, CI.getFileSystemOpts()));
+        Unit = LoadedUnit.get();
+
+        // Cache the resulting ASTUnit.
+        if (Unit)
```

Comment:
We should add an else branch here with some diagnostics, because I think if `Unit` is null then something went wrong with the load (e.g. wrong .ast format, etc).